python/ray/util/client/common.py

```diff
@@ -217,61 +247,89 @@ def __init__(self,
         self.actor_ref = actor_ref
         self._dir: Optional[List[str]] = None
         if actor_class is not None:
-            self._dir = list(
-                dict(
-                    inspect.getmembers(actor_class.actor_cls,
-                                       is_function_or_method)).keys())
+            self._method_num_returns = {}
```

Comment:
I suggest splitting this PR into two: One for the async submission, and one (smaller one) for the num_returns fix and the `data = ray.worker._dumps_from_client(self.actor_cls)` change above.

Maybe you have ideas how it could be split even more.

It is quite possible that this PR will cause problems further down the line (just by the nature of the change), so having it chunked into smaller pieces will make it much easier to debug problems.

Reply:
Good suggestion, since we need to wait for other changes to merge too. I'm thinking of a preliminary change that allows 
1.  Client{Object,Actor}Ref to take futures
2. Store num_returns & function signatures, check them synchronously
(2) is needed by the async change for the (desirable) behavior of raise `TypeError` synchronously when functions are called with incorrect arguments. They should have no / minimal conflict with other PRs.

Reply:
Splitted out a (majority) part of the code to #18677, which should not have significant behavior change.